### PR TITLE
Fix bugs on attachments

### PR DESCRIPTION
**Pull request by Birmania**
-Double in attachments (as the first FILE is already processed by the original plugin)
-"+" operator  replaced  by array_unique(array_merge...)) as "+" is an union on keys and not values.
-"dup_tag" replaced by "sg_field_tag"
-$components['attachments'] is an array, use the addition operator
-"apply_filters" instead of "apply_fitlers"

### CODE REVIEW

**Comment by Birmania** [comment]
Note that I discovered a new bug which is not solved in this PR : 
In the "filter_wpcf7_validate" method, when there is the "@since 2.1.5 fix validation of non-toggled checkox/radio" code. In fact it only verify the first element of a tag and not the tabbed ones. That imply some strange behaviors like : 

If there is a radio which is neither selected in the first and second tab, the validation only occurs on  the first tab.
If there is a input file which is only set in the second tab, it will be lost as the "validate" did not occurs.

**Comment by Birmania** [comment]
Hi @aurovrata 

> Note that I discovered a new bug which is not solved in this PR : 

O.K, I think that I have a fix ! I
I clean it a bit and send you the PR, let's count 24 hours.

**Comment by Birmania** [comment]
Hi @aurovrata 

Sorry for the time it took but I tried three different design and was only satisfied by the simplicity of this last one.
This is a major change on consolidation phase as it will no more try to guess the fields on server-side. It will use a list built thanks to JS.

This will solve problem with Checkbox/Radio that are not sent when empty.
It may be my point of view but I feel server-side methods more simple this way.

**Comment by Birmania** [comment]
O.K, I think I didn't test everything. It was your backend consolidation that set the values for posted_data and so mail sent values.
I will modify my PR tomorrow to keep the consolidation (based on cf7sg_fields) and validate using the posted data hierarchy (as you did before).

Do not merge now please. :)

**Comment by Birmania** [comment]
Hi back @aurovrata,

Sorry for the time it take, I just finished the PR and validate it with : 
- Validation of Singular fields
- Validation of Fields in tabs
- Validation of Fields in rows
- Validation of Fields in tabs+rows
- Sending of form with all the previous listed fields + tabbed Files/Attachments

Everything seems O.K !
Yours faithfully,
Birmania

**Comment by aurovrata** [comment]
Hey Birmania,  
Sorry for the late reply, I was travelling for work and could not look at your messages.
thanks for the PR.  Let me go through it and get back to you.

**Comment by Birmania** [comment]
Hi aurovrata,

No problem, hope your travel was pleasant even for working.

This PR rework a bit the form post consolidation philosophy but it should cover/fix many special cases.

File: public/class-cf7-grid-layout-public.php [hunk 1]
```diff
@@ -1212,15 +1211,14 @@ public function wpcf7_mail_components($components, $cf7form, $cf7mail) {
            case 'tab':
            case 'table':
            case 'both':
-             $regex = '/^'.preg_quote($tag['name']);
              // Search for $_FILES where name match a tabbed file field
-             $submitted_fields = preg_grep($regex.'$/', $uploaded_files);
```
Comment (aurovrata) [inline public/class-cf7-grid-layout-public.php:1217]: not sure why you deleted this line.  the first tab fields use the original field names, subsequent tabs use \<field-name\>_tab-1, \<field-name\>_tab-2 ...
for rows (in tables) again, the same logic, the first row uses the naked field name, and subsequent rows have their field names affixed with the row suffix: \<field-name\>_row-1, \<field-name\>_row-2...

for tables within tabs, the first first row in the first tab is the naked field name, then the next row is \<field-name\>_row-1 and so on, In the 2nd tab you have the first row as \<field-name\>_tab-1, and the subsequent row as \<field-name\>_tab-1_row-1, \<field-name\>_tab-1_row-2 and so on.

Hence we need to still search for the naked field name.

**Comment by aurovrata** [comment]
> -"+" operator replaced by array_unique(array_merge...)) as "+" is an union on keys and not values.

yes, you are right except that you have not taken into account the way [preg_grep](https://secure.php.net/manual/en/function.preg-grep.php) works. It,

> Returns an array indexed using the keys from the input array. 

and this is the reason why it works,  since I am taking the keys to build an array of values, preg_grep extracts these as say,

( [1]=> \<matched-key\>,
[3] =>  \<matched-key\>,)

in the next preg_grep results, the values are:

( [0]=> \<matched-key\>,
[5] =>  \<matched-key\>,)

and therefore adding the above 2 arrays using a '+' operator results in,

( [1]=> \<matched-key\>,
[3] => \<matched-key\>,
[0]=> \<matched-key\>,
[5] => \<matched-key\>,)

I guess the code is confusign to read, and would likely be clearer using your expanded code, but given that I have used this in multiple places in the plugin code as well as making it more concise, I prefer to stick to this code structure.

File: public/class-cf7-grid-layout-public.php (hunk 1 above)
Comment (Birmania) [reply public/class-cf7-grid-layout-public.php:1217]: In fact, processing of naked upload files is already done by the CF7 validate methods. ("validate" in "submission.php" and "wpcf7_file_validation_filter" in "file.php")
Thus, we only need to call validation on duplicated fields (tabs, rows) in order to fill the uploaded_files array with missing FILES.

**Comment by Birmania** [comment]
Hi Aurovrata,

Thanks for this info, I clearly missed it !
You are right, keep your writing style with '+', it is a good visual marker.

**Comment by aurovrata** [comment]
Dear Birmania,

so after quite a protracted period of coding/testing I finally came to the same conclusion as yours, that a tracking of tab/table/toggled section fields was required in order to stop guessing how many such fields were submitted in order to properly solve the problem of file and checkbox fields validation.

I was already partially tracking tab and table fields, and now I added a hidden field for each table/tab sections to track the number of rows/panels added.  For toggle fields 

For toggled sections, the plugin was already tracking their usage, but this was not integrated yet with the validation.  This is now done.

Could you get some time to test to see if it works with your installation/forms as per your expectations?

## Moving forward.

I appreciate your enthusiasm for helping out with improving/fixing this plugin, however, I realised that due to the lack of proper code documentation, it is difficult to make meaningful contributions to what can be complex data manipulations.

Hence, if you are keen to continue to collaborate, I propose to start documentation various classes/php files so as to make it clearer for you to understand.

I would also propose that before you embark on further code modifications, that we have have an intial discussion by raising an issue in order for me to give you specific guidance and therefore ensuring that your efforts are not in vain.

What do you think?

**Comment by Birmania** [comment]
Hi @aurovrata,

Happy new year to you and your family.
Sincerely sorry for the time I take to answer... I am totally busy with my "traditional" work...

I agree with all you said ! I need to create issues first to talk about the subjects.

As for the new versions, the behavior seems perfect, validation is going well. Good job. ;)

**Comment by aurovrata** [comment]
HI @Birmania great to read you, wishing you a very good 2019 :) 

I have a few ideas for improving the plugin and I am going to draw up a wiki page with some of these so we can maybe collaborate on some of the functional development.

**Comment by aurovrata** [comment]
Here is a [roadmap](https://github.com/aurovrata/cf7-grid-layout/wiki/Roadmap) requirements page.

Take a look and let me know if there is anything that might of interest to you.

**Pull request closed without merging**